fdbclient/Tracing.actor.cpp

```diff
@@ -167,147 +168,48 @@ ACTOR Future<Void> traceLog(int* pendingMessages, bool* sendError) {
-			serialize_value(link.traceID.second(), request, 0xcf); // trace id
-			serialize_value(link.spanID, request, 0xcf); // spanid
-		}
-	}
-
-	// Writes the given vector of linked SpanContext's to the request. If the vector is
-	// empty, the request is not modified.
-	inline void serialize_vector(const SmallVectorRef<SpanEventRef>& vec, TraceRequest& request) {
-		int size = vec.size();
-		if (size <= 15) {
-			request.write_byte(static_cast<uint8_t>(size) | 0b10010000);
-		} else if (size <= 65535) {
-			request.write_byte(0xdc);
-			request.write_byte(reinterpret_cast<const uint8_t*>(&size)[1]);
-			request.write_byte(reinterpret_cast<const uint8_t*>(&size)[0]);
-		} else {
-			TraceEvent(SevWarn, "TracingSpanSerializeVector").detail("Failed to MessagePack encode large vector", size);
-			ASSERT_WE_THINK(false);
-		}
-
-		for (const auto& event : vec) {
-			serialize_string(event.name.toString(), request); // event name
-			serialize_value(event.time, request, 0xcb); // event time
-			serialize_vector(event.attributes, request);
-		}
-	}
-
-	inline void serialize_vector(const SmallVectorRef<KeyValueRef>& vals, TraceRequest& request) {
-		int size = vals.size();
-		if (size <= 15) {
-			// N.B. We're actually writing this out as a fixmap here in messagepack format!
-			// fixmap	1000xxxx	0x80 - 0x8f
-			request.write_byte(static_cast<uint8_t>(size) | 0b10000000);
-		} else {
-			TraceEvent(SevWarn, "TracingSpanSerializeVector").detail("Failed to MessagePack encode large vector", size);
-			ASSERT_WE_THINK(false);
-		}
+		serialize_value(span.status, buf, 0xcc);
 
-		for (const auto& kv : vals) {
-			serialize_string(kv.key.toString(), request);
-			serialize_string(kv.value.toString(), request);
-		}
-	}
-
-	template <class Map>
-	inline void serialize_map(const Map& map, TraceRequest& request) {
-		int size = map.size();
+		// Links
+		std::function<void(const SpanContext&, MsgpackBuffer&)> f_packLinks = [](const SpanContext& ctx,
+		                                                                         MsgpackBuffer& buf) {
+			serialize_value(ctx.traceID.first(), buf, 0xcf); // trace id
+			serialize_value(ctx.traceID.second(), buf, 0xcf); // trace id
+			serialize_value(ctx.spanID, buf, 0xcf); // spanid
+		};
+		serialize_vector(span.links, buf, f_packLinks);
 
-		if (size <= 15) {
-			request.write_byte(static_cast<uint8_t>(size) | 0b10000000);
-		} else {
-			TraceEvent(SevWarn, "TracingSpanSerializeMap").detail("Failed to MessagePack encode large map", size);
-			ASSERT_WE_THINK(false);
-		}
+		// Events
+		std::function<void(const SpanEventRef&, MsgpackBuffer&)> f_packEvents = [](const SpanEventRef& event,
+		                                                                           MsgpackBuffer& buf) {
+			serialize_string(event.name.toString(), buf); // event name
+			serialize_value(event.time, buf, 0xcb); // event time
+			std::function<void(const KeyValueRef&, MsgpackBuffer&)> f_packAttributes = [](const KeyValueRef& kv,
```

Comment:
I worry a bit about the performance of all these `std::function` usages on the very hot path of span serialization. @sfc-gh-mpilman do you have any thoughts on the potential performance hit here?

Reply:
I can make a templatated version that can take a function ptr or a functor too. Something like
`
template<typname T, typname F> serialize_vector(const std::vector<T>&, MsgpackBuffer&, F f);
`
That way with a function ptr, we don't deal with the overhead of `std::function` as much?

Reply:
Did you revert the span tracing so that it uses the existing `serialize_vector` functions?

Since we are planning to be using tracing in production soon, any potential performance hit in this path will need to be tested to ensure it doesn't cause a performance regression.

Reply:
No, I kept those distinct since it doesn't use the array format of msg pack, but rather `fixmap` so it's a different encoding